// @ezeep/ezeep-js/dist/collection/components/ezp-label/ezp-label.js
import { Component, Host, Prop, h } from '@stencil/core';
export class EzpLabel {
  constructor() {
    /**
     *
     * Properties
     *
     */
    /** Description... */
    this.ellipsis = false;
    /** Description... */
    this.level = 'secondary';
    /** Description... */
    this.noWrap = false;
    /** Description... */
    this.text = 'Label';
    /** Description... */
    this.weight = 'soft';
  }
  /**
   *
   * Render method
   *
   */
  render() {
    return (h(Host, { class: `${this.level} ${this.weight} ${this.ellipsis ? 'ellipsis' : ''} ${this.noWrap ? 'no-wrap' : ''}` },
      h("div", { id: "text" }, this.text)));
  }
  static get is() { return "ezp-label"; }
  static get encapsulation() { return "shadow"; }
  static get originalStyleUrls() { return {
    "$": ["ezp-label.scss"]
  }; }
  static get styleUrls() { return {
    "$": ["ezp-label.css"]
  }; }
  static get properties() { return {
    "ellipsis": {
      "type": "boolean",
      "mutable": false,
      "complexType": {
        "original": "boolean",
        "resolved": "boolean",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": "Description..."
      },
      "attribute": "ellipsis",
      "reflect": false,
      "defaultValue": "false"
    },
    "level": {
      "type": "string",
      "mutable": false,
      "complexType": {
        "original": "LabelLevelTypes",
        "resolved": "\"primary\" | \"secondary\" | \"tertiary\"",
        "references": {
          "LabelLevelTypes": {
            "location": "import",
            "path": "../../shared/types"
          }
        }
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": "Description..."
      },
      "attribute": "level",
      "reflect": false,
      "defaultValue": "'secondary'"
    },
    "noWrap": {
      "type": "boolean",
      "mutable": false,
      "complexType": {
        "original": "boolean",
        "resolved": "boolean",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": "Description..."
      },
      "attribute": "no-wrap",
      "reflect": false,
      "defaultValue": "false"
    },
    "text": {
      "type": "string",
      "mutable": false,
      "complexType": {
        "original": "string",
        "resolved": "string",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": "Description..."
      },
      "attribute": "text",
      "reflect": false,
      "defaultValue": "'Label'"
    },
    "weight": {
      "type": "string",
      "mutable": false,
      "complexType": {
        "original": "WeightTypes",
        "resolved": "\"heavy\" | \"soft\" | \"strong\"",
        "references": {
          "WeightTypes": {
            "location": "import",
            "path": "../../shared/types"
          }
        }
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": "Description..."
      },
      "attribute": "weight",
      "reflect": false,
      "defaultValue": "'soft'"
    }
  }; }
}
