// @ezeep/ezeep-js/dist/collection/components/ezp-text-button/ezp-text-button.js
import { Component, Host, Prop, h } from '@stencil/core';
export class EzpTextButton {
  constructor() {
    /**
     *
     * Properties
     *
     */
    /** Description... */
    this.blank = false;
    /** Description... */
    this.disabled = false;
    /** Description... */
    this.level = 'primary';
  }
  /**
   *
   * Render method
   *
   */
  render() {
    const TagType = this.type !== undefined ? 'button' : 'a';
    const attributes = this.type !== undefined
      ? {
        type: this.type,
        disabled: this.disabled,
      }
      : {
        href: this.href,
        target: this.blank ? '_blank' : '_self',
      };
    return (h(Host, { class: `${this.level}` },
      h(TagType, Object.assign({ class: "button" }, attributes),
        h("ezp-label", { weight: "heavy", text: this.label }))));
  }
  static get is() { return "ezp-text-button"; }
  static get encapsulation() { return "shadow"; }
  static get originalStyleUrls() { return {
    "$": ["ezp-text-button.scss"]
  }; }
  static get styleUrls() { return {
    "$": ["ezp-text-button.css"]
  }; }
  static get properties() { return {
    "blank": {
      "type": "boolean",
      "mutable": false,
      "complexType": {
        "original": "boolean",
        "resolved": "boolean",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": "Description..."
      },
      "attribute": "blank",
      "reflect": false,
      "defaultValue": "false"
    },
    "disabled": {
      "type": "boolean",
      "mutable": false,
      "complexType": {
        "original": "boolean",
        "resolved": "boolean",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": "Description..."
      },
      "attribute": "disabled",
      "reflect": false,
      "defaultValue": "false"
    },
    "href": {
      "type": "string",
      "mutable": false,
      "complexType": {
        "original": "string",
        "resolved": "string",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": "Description..."
      },
      "attribute": "href",
      "reflect": false
    },
    "level": {
      "type": "string",
      "mutable": false,
      "complexType": {
        "original": "TextButtonLevelTypes",
        "resolved": "\"primary\" | \"secondary\" | \"tertiary\"",
        "references": {
          "TextButtonLevelTypes": {
            "location": "import",
            "path": "../../shared/types"
          }
        }
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": "Description..."
      },
      "attribute": "level",
      "reflect": false,
      "defaultValue": "'primary'"
    },
    "label": {
      "type": "string",
      "mutable": false,
      "complexType": {
        "original": "string",
        "resolved": "string",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": "Description..."
      },
      "attribute": "label",
      "reflect": false
    },
    "type": {
      "type": "string",
      "mutable": false,
      "complexType": {
        "original": "TextButtonTypeTypes",
        "resolved": "\"button\"",
        "references": {
          "TextButtonTypeTypes": {
            "location": "import",
            "path": "../../shared/types"
          }
        }
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": "Description..."
      },
      "attribute": "type",
      "reflect": false
    }
  }; }
}
